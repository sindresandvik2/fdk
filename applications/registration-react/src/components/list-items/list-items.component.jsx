import React from 'react';
import PropTypes from 'prop-types';
import orderBy from 'lodash/orderBy';
import _ from 'lodash';
import { withState, withHandlers, compose } from 'recompose';

import localization from '../../utils/localization';
import SortButtons from '../sort-button/sort-button.component';
import { ListItem } from './list-item/list-item.component';
import getTranslateText from '../../utils/translateText';
import './list-items.scss';

const renderItems = (
  catalogId,
  items,
  itemTitleField,
  sortField,
  sortType,
  prefixPath,
  defaultEmptyListText
) => {
  if (items) {
    let sortedItems = items;
    if (sortField === 'title') {
      // order by title and ignore case
      sortedItems = orderBy(
        items,
        [
          item => {
            if (_.get(item, itemTitleField)) {
              const retTitle =
                getTranslateText(_.get(item, itemTitleField)) ||
                _.get(item, itemTitleField);
              return retTitle.toLowerCase();
            }
            return null;
          }
        ],
        [sortType]
      );
    } else if (sortField === 'registrationStatus') {
      sortedItems = orderBy(items, 'registrationStatus', [sortType]);
    }

    return sortedItems.map(item => {
      const title =
        getTranslateText(_.get(item, itemTitleField)) ||
        _.get(item, itemTitleField);
      return (
        <ListItem
          key={item.id}
          title={typeof title === 'object' ? null : title}
          status={_.get(item, 'registrationStatus')}
          path={`${prefixPath}/${item.id}`}
        />
      );
    });
  }
  return (
    <div className="fdk-list-item d-flex">
      <span className="fdk-text-size-small fdk-color2">
        {defaultEmptyListText}
      </span>
    </div>
  );
};

export const ListItems = props => {
  const {
    catalogId,
    items,
    itemTitleField,
    sortField,
    sortType,
    onSortField,
    prefixPath,
    defaultEmptyListText
  } = props;
  return (
    <div>
      <div className="fdk-list-header d-flex">
        <div className="d-flex align-items-center w-75">
          <span className="header-item mr-1">{localization.title}</span>
          <SortButtons
            field="title"
            sortField={sortField}
            sortType={sortType}
            onSortField={onSortField}
          />
        </div>

        <div className="d-flex align-items-center w-25">
          <span className="header-item mr-1">
            {localization.listItems.header.status}
          </span>
          <SortButtons
            field="registrationStatus"
            sortField={sortField}
            sortType={sortType}
            onSortField={onSortField}
          />
        </div>
      </div>
      {renderItems(
        catalogId,
        items,
        itemTitleField,
        sortField,
        sortType,
        prefixPath,
        defaultEmptyListText
      )}
    </div>
  );
};

ListItems.defaultProps = {
  catalogId: null,
  items: null,
  itemTitleField: ['title'],
  sortField: null,
  sortType: null,
  onSortField: null,
  prefixPath: null,
  defaultEmptyListText: null
};

ListItems.propTypes = {
  catalogId: PropTypes.string.isRequired,
  items: PropTypes.array,
  itemTitleField: PropTypes.array,
  sortField: PropTypes.string,
  sortType: PropTypes.string,
  onSortField: PropTypes.func,
  prefixPath: PropTypes.string,
  defaultEmptyListText: PropTypes.string
};

const enhance = compose(
  withState('sortField', 'setSortField', ''),
  withState('sortType', 'setSortType', ''),
  withHandlers({
    onSortField: props => (field, type) => {
      props.setSortField(field);
      props.setSortType(type);
    }
  })
);

export default enhance(ListItems);
