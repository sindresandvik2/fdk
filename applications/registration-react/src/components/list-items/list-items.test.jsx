import React from 'react';
import { shallow } from 'enzyme';
import { ListItems } from './list-items.component';

let defaultProps;
let wrapper;

beforeEach(() => {
  const items = {
    items: [
      {
        id: 1,
        title: {
          nb: 'Test item'
        },
        registrationStatus: 'DRAFT'
      },
      {
        id: 2,
        title: {
          nb: 'Test item'
        },
        registrationStatus: 'PUBLISH'
      }
    ]
  };
  const onSort = jest.fn();
  defaultProps = {
    catalogId: '123',
    items: items.items,
    sortField: 'title',
    sortType: 'asc',
    onSortField: onSort,
    prefixPath: '/catalogs/123/datasets',
    defaultEmptyListText: 'Ingen beskrivelser er registrert'
  };
  wrapper = shallow(<ListItems {...defaultProps} />);
});

test('should render ListItem correctly with props', () => {
  expect(wrapper).toMatchSnapshot();
});
