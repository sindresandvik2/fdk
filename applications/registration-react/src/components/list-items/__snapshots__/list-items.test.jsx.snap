// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ListItem correctly with props 1`] = `
<div>
  <div
    className="fdk-list-header d-flex"
  >
    <div
      className="d-flex align-items-center w-75"
    >
      <span
        className="header-item mr-1"
      >
        Tittel
      </span>
      <SortButtons
        field="title"
        onSortField={[MockFunction]}
        sortField="title"
        sortType="asc"
      />
    </div>
    <div
      className="d-flex align-items-center w-25"
    >
      <span
        className="header-item mr-1"
      >
        Status
      </span>
      <SortButtons
        field="registrationStatus"
        onSortField={[MockFunction]}
        sortField="title"
        sortType="asc"
      />
    </div>
  </div>
  <ListItem
    key="1"
    path="/catalogs/123/datasets/1"
    status="DRAFT"
    title="Test item"
  />
  <ListItem
    key="2"
    path="/catalogs/123/datasets/2"
    status="PUBLISH"
    title="Test item"
  />
</div>
`;
